taichi/backends/opengl/codegen_opengl.cpp

```diff
@@ -627,6 +642,39 @@ class KernelGen : public IRVisitor {
   void visit(AtomicOpStmt *stmt) override {
     TI_ASSERT(stmt->width() == 1);
     auto dt = stmt->dest->element_type().ptr_removed();
+
+    auto val_name = stmt->val->short_name();
+
+    if (stmt->is_reduction &&
+        (dt->is_primitive(PrimitiveTypeID::f32) ||
+         dt->is_primitive(PrimitiveTypeID::i32) ||
+         dt->is_primitive(PrimitiveTypeID::u32)) &&
+        (stmt->op_type == AtomicOpType::add ||
+         stmt->op_type == AtomicOpType::sub ||
+         stmt->op_type == AtomicOpType::min ||
+         stmt->op_type == AtomicOpType::max)) {
+      used.reduction = true;
+      val_name += "_reduction";
+
+      auto op_name = "";
+
+      if (stmt->op_type == AtomicOpType::sub ||
+          stmt->op_type == AtomicOpType::add) {
+        op_name = "add";
+      } else if (stmt->op_type == AtomicOpType::max) {
+        op_name = "max";
+      } else if (stmt->op_type == AtomicOpType::min) {
+        op_name = "min";
+      }
+
+      emit("{} {} = reduction_workgroup_{}_{}({});",
+           opengl_data_type_name(stmt->val->element_type()), val_name, op_name,
+           opengl_data_type_name(stmt->val->element_type()),
+           stmt->val->short_name());
+
+      emit("if (gl_LocalInvocationIndex == 0)");
+    }
+
```

Comment:
nit: I think it's better to emit an additional pair of `{}` between L678-707. That way if we ever emit more than one line, it won't suddenly break  when workgroup reduction is enabled.